_posts/2020-09-07-spring-boot-health-check.md

```diff
@@ -0,0 +1,365 @@
+
+### Aggregating Health Indicators
+The aggregation is done by an implementation of StatusHealthAggregator which aggregates the statuses from each health indicator into a single overall status.
+
+Spring Boot auto-configures an instance of SimpleHealthAggregator. We can provide our implementation of StatusHealthAggregator to supersede the default behavior. 
+
+We can also disable a particular health indicator using application properties :
+```
+management.health.mongo.enabled=false
+```
+
+## Checking the Health of APIs with Custom Health Indicators
+Predefined Health Indicators do not cover all use cases of Health Check. For example, if our API is dependent on any external service, we might like to know if the external service is available. 
+
+In our example, we are using an external service for shortening the URLs. We will monitor the availability of this service by building our health indicator. 
+
+Creating custom health indicators is done in two steps:
+1. Implement the HealthIndicator interface and override the health method.
+2. Register the health indicator class as Spring Bean by adding the `component` annotation.
+
+Our custom Health Indicator for UrlShortener Service looks like this:
+
+```java
+@Component
+@Slf4j
+public class UrlShortenerServiceHealthIndicator implements HealthIndicator, HealthContributor {
+
+  private static final String URL = "https://cleanuri.com/api/v1/shorten";
+
+  @Override
+  public Health health() {
+    // check if url shortener service url is reachable
+        try {
+            URL url = new URL(URL);
+            int port = url.getPort();
+            if (port == -1) {
+                port = url.getDefaultPort();
+            }
+
+            try (Socket socket = new Socket(url.getHost(), port)) {
+            } catch (IOException e) {
+                log.warn("Failed to connect to : {}", URL);
+                return Health.down().withDetail("error", e.getMessage()).build();
+            }
+        } catch (MalformedURLException e1) {
+            log.warn("Invalid URL: {}",URL);
+            return Health.down().withDetail("error", e1.getMessage()).build();
+        }
+
+        return Health.up().build();
+  }
+```
+In this class, we return the status as "UP" if the URL is reachable, otherwise we return the "DOWN" status with an error message.
+
+We can further increase the accuracy of the health checks by checking specific resources on a per API basis. 
+
+In the previous section, we added three APIs to our application. It will be very useful to see the health of the APIs as part of our health check. We will do this by building a specific health indicator for our API. The health status of our API will be "UP" only if the resources like the database and external service on which it depends are available:
+
+A snippet of the `FetchUsersAPIHealthIndicator` is given below:
+```java
+@Component("FetchUsersAPI")
```

Comment:
I don't understand the intent of this code snippet. I can see no reference to "fetching users" in the code. What to you want to show with this code snippet? How to manually aggregate multiple health indicators to a "deep check"? In that case, please make this a separate section with a separate section header.